plugins/amdgpu/amdgpu_plugin.c

```diff
@@ -1862,3 +1941,233 @@ int amdgpu_plugin_resume_devices_late(int target_pid)
+	int ret = 0;
+	int offset = 0;
+	void *buffer = NULL;
+
+	ret = init_dev(thread_data->minor, &h_dev, &max_copy_size);
+	if (ret < 0) {
+		goto err;
+	}
+
+	for (int i = 0; i < restore_cmd->cmd_head.entry_num; i++) {
+		if (restore_cmd->entries[i].gpu_id == thread_data->gpu_id) {
+			total_bo_size += restore_cmd->entries[i].size;
+
+			if (restore_cmd->entries[i].size > max_bo_size)
+				max_bo_size = restore_cmd->entries[i].size;
+		}
+	}
+
+	buffer_size = kfd_max_buffer_size > 0 ? min(kfd_max_buffer_size, max_bo_size) : max_bo_size;
+
+	bo_contents_fp = get_bo_contents_fp(restore_cmd->cmd_head.id, thread_data->gpu_id, total_bo_size);
+	if (bo_contents_fp == NULL) {
+		ret = -1;
+		goto err_sdma;
+	}
+	offset = ftell(bo_contents_fp);
+
+	posix_memalign(&buffer, sysconf(_SC_PAGE_SIZE), buffer_size);
+	if (!buffer) {
+		pr_perror("Failed to alloc aligned memory. Consider setting KFD_MAX_BUFFER_SIZE.");
+		ret = -ENOMEM;
+		goto err_sdma;
+	}
+
+	for (int i = 0; i < restore_cmd->cmd_head.entry_num; i++) {
+		if (restore_cmd->entries[i].gpu_id != thread_data->gpu_id)
+			continue;
+
+		entry = &restore_cmd->entries[i];
+		fseek(bo_contents_fp, entry->read_offset + offset, SEEK_SET);
+		ret = sdma_copy_bo_helper(entry->size, restore_cmd->fds_write[entry->write_id], bo_contents_fp, buffer,
+					  buffer_size, h_dev, max_copy_size, SDMA_OP_VRAM_WRITE);
+		if (ret) {
+			pr_err("Failed to fill the BO using sDMA: bo_buckets[%d]\n", i);
+			goto err_sdma;
+		}
+	}
+
+err_sdma:
+	if (bo_contents_fp)
+		fclose(bo_contents_fp);
+	if (buffer)
+		xfree(buffer);
+	amdgpu_device_deinitialize(h_dev);
+err:
+	thread_data->ret = ret;
+	return NULL;
+}
+
+void *restore_device_parallel_worker(void *arg)
+{
+	while (1) {
+		parallel_restore_cmd restore_cmd = { 0 };
+		struct parallel_thread_data *thread_datas = NULL;
+		int ret;
+
+		pr_debug("Begin to recv parallel restore cmd\n");
+		ret = recv_parallel_restore_cmd(&restore_cmd);
+		if (ret) {
+			if (ret == 1) {
+				*(int *)arg = 0;
+				goto exit;
+			}
+			goto err;
+		}
+
+		thread_datas = xzalloc(sizeof(*thread_datas) * restore_cmd.cmd_head.gpu_num);
+		if (!thread_datas) {
+			ret = -ENOMEM;
+			goto err;
+		}
+
+		for (int i = 0; i < restore_cmd.cmd_head.gpu_num; i++) {
+			thread_datas[i].gpu_id = restore_cmd.gpu_ids[i].gpu_id;
+			thread_datas[i].minor = restore_cmd.gpu_ids[i].minor;
+			thread_datas[i].restore_cmd = &restore_cmd;
+
+			ret = pthread_create(&thread_datas[i].thread, NULL, parallel_restore_bo_contents,
+					     (void *)&thread_datas[i]);
+			if (ret) {
+				pr_err("Failed to create thread[%i] ret:%d\n", i, ret);
+				goto err;
+			}
+		}
+
+		for (int i = 0; i < restore_cmd.cmd_head.gpu_num; i++) {
+			pthread_join(thread_datas[i].thread, NULL);
+
+			pr_info("Thread[0x%x] finished ret:%d\n", thread_datas[i].gpu_id, thread_datas[i].ret);
+
+			if (thread_datas[i].ret) {
+				ret = thread_datas[i].ret;
+				goto err;
+			}
+		}
+
+err:
+		if (thread_datas)
+			xfree(thread_datas);
```

Comment:
thread_datas can't be freed until all threads have exited. In the previous loop, you go to the err label when one thread fails, so some other threads can be still running.